xapian-core/weight/tfidfweight.cc

```diff
@@ -36,85 +36,68 @@ using namespace std;
 
 namespace Xapian {
 
-TfIdfWeight::TfIdfWeight(const std::string& normals)
-    : TfIdfWeight::TfIdfWeight(normals, 0.2, 1.0) {}
-
-TfIdfWeight::TfIdfWeight(const std::string& normals, double slope, double delta)
-    : param_slope(slope), param_delta(delta)
+TfIdfWeight::wdf_norm
+TfIdfWeight::decode_wdf_norm(char c)
 {
-    if (normals.length() != 3 ||
-	!strchr("nbslPL", normals[0]) ||
-	!strchr("ntpfsP", normals[1]) ||
-	!strchr("n", normals[2]))
-	throw Xapian::InvalidArgumentError("Normalization string is invalid");
-    if (param_slope <= 0)
-	throw Xapian::InvalidArgumentError("Parameter slope is invalid");
-    if (param_delta <= 0)
-	throw Xapian::InvalidArgumentError("Parameter delta is invalid");
-    if (normals[1] != 'n') {
-	need_stat(TERMFREQ);
-	need_stat(COLLECTION_SIZE);
-    }
-    need_stat(WDF);
-    need_stat(WDF_MAX);
-    need_stat(WQF);
-    if (normals[0] == 'P' || normals[1] == 'P') {
-	need_stat(AVERAGE_LENGTH);
-	need_stat(DOC_LENGTH);
-	need_stat(DOC_LENGTH_MIN);
-    }
-    if (normals[0] == 'L') {
-	need_stat(DOC_LENGTH);
-	need_stat(DOC_LENGTH_MIN);
-	need_stat(DOC_LENGTH_MAX);
-	need_stat(UNIQUE_TERMS);
-    }
-    switch (normals[0]) {
+    switch (c) {
 	case 'b':
-	    wdf_norm_ = wdf_norm::BOOLEAN;
-	    break;
+	    return wdf_norm::BOOLEAN;
 	case 's':
-	    wdf_norm_ = wdf_norm::SQUARE;
-	    break;
+	    return wdf_norm::SQUARE;
 	case 'l':
-	    wdf_norm_ = wdf_norm::LOG;
-	    break;
+	    return wdf_norm::LOG;
 	case 'P':
-	    wdf_norm_ = wdf_norm::PIVOTED;
-	    break;
+	    return wdf_norm::PIVOTED;
 	case 'L':
-	    wdf_norm_ = wdf_norm::LOG_AVERAGE;
-	    break;
+	    return wdf_norm::LOG_AVERAGE;
+	case 'n':
+	    return wdf_norm::NONE;
 	default:
-	    wdf_norm_ = wdf_norm::NONE;
+	    throw Xapian::InvalidArgumentError
+		  ("Normalization string is invalid");
     }
-    switch (normals[1]) {
+}
+
+TfIdfWeight::idf_norm
+TfIdfWeight::decode_idf_norm(char c)
+{
+    switch (c) {
 	case 'n':
-	    idf_norm_ = idf_norm::NONE;
-	    break;
+	    return idf_norm::NONE;
 	case 's':
-	    idf_norm_ = idf_norm::SQUARE;
-	    break;
+	    return idf_norm::SQUARE;
 	case 'f':
-	    idf_norm_ = idf_norm::FREQ;
-	    break;
+	    return idf_norm::FREQ;
 	case 'P':
-	    idf_norm_ = idf_norm::PIVOTED;
-	    break;
+	    return idf_norm::PIVOTED;
 	case 'p':
-	    idf_norm_ = idf_norm::PROB;
-	    break;
+	    return idf_norm::PROB;
+	case 't':
+	    return idf_norm::TFIDF;
 	default:
-	    idf_norm_ = idf_norm::TFIDF;
+	    throw Xapian::InvalidArgumentError
+		  ("Normalization string is invalid");
     }
-    wt_norm_ = wt_norm::NONE;
 }
```

Comment:
Since all the other cases `return`, you could handle the default case outside the `switch` which would avoid the wrapping.

Reply:
Done.